Helm 3: "required" template function breaks linter

Output of `helm version`:
```
$ helm version
version.BuildInfo{Version:"v3.0.0-alpha.1", GitCommit:"b9a54967f838723fe241172a6b94d18caf8bcdca", GitTreeState:"clean"}
```

Output of `kubectl version`:
N/A

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
N/A

If a chart uses the `required` filter with no default value running `helm lint` over the chart will fail, this does not happen in helm 2.

An example chart that will reproduce this issue is as follows:

**Chart.yaml**
```yaml
apiVersion: v1
name: test
description: A Helm chart for Kubernetes

# A chart can be either an 'application' or a 'library' chart.
#
# Application charts are a collection of templates that can be packaged into versioned archives
# to be deployed.
#
# Library charts provide useful utilities or functions for the chart developer. They're included as
# a dependency of application charts to inject those utilities and functions into the rendering
# pipeline. Library charts do not define any templates and therefore cannot be deployed.
type: application

# This is the chart version. This version number should be incremented each time you make changes
# to the chart and its templates, including the app version.
version: 0.1.0

# This is the version number of the application being deployed. This version number should be
# incremented each time you make changes to the application.
appVersion: 1.16.0
```
**values.yaml**
```yaml
---
``` 
**templates/foo.yaml**
```yaml
---
foo: {{ .Values.unset | required "uset is required!" }}
```

When running `helm lint` on this chart you get the following error:
```
Error: 1 chart(s) linted, 1 chart(s) failed
	%!s(<nil>)
	[INFO] Chart.yaml: icon is recommended
	[ERROR] templates/: render error at (test/templates/foo.yaml:3): 'uset is required!'
```

When running the same command using helm 2 the chart is valid
```
$ helm2 version --client
Client: &version.Version{SemVer:"v2.14.1", GitCommit:"5270352a09c7e8b6e8c9593002a73535276507c0", GitTreeState:"clean"}

$ helm2 lint .
2019/06/13 16:00:11 [INFO] Missing required value: uset is required!
==> Linting .
[INFO] Chart.yaml: icon is recommended

1 chart(s) linted, no failures
```

I believe this maybe a regression of #4747 and #4221.

--- a/pkg/engine/engine.go
+++ b/pkg/engine/engine.go
@@ -18,6 +18,7 @@ package engine
 
 import (
 	"fmt"
+	"log"
 	"path"
 	"path/filepath"
 	"sort"
@@ -35,6 +36,8 @@ type Engine struct {
 	// If strict is enabled, template rendering will fail if a template references
 	// a value that was not passed in.
 	Strict bool
+	// In LintMode, some 'required' template values may be missing, so don't fail
+	LintMode bool
 }
 
 // Render takes a chart, optional values, and value overrides, and attempts to render the Go templates.
@@ -114,6 +117,29 @@ func (e Engine) initFunMap(t *template.Template, referenceTpls map[string]render
 		}
 		return result[templateName.(string)], nil
 	}
+
+	// Add the `required` function here so we can use lintMode
+	funcMap["required"] = func(warn string, val interface{}) (interface{}, error) {
+		if val == nil {
+			if e.LintMode {
+				// Don't fail on missing required values when linting
+				log.Printf("[INFO] Missing required value: %s", warn)
+				return "", nil
+			}
+			return val, errors.Errorf(warn)
+		} else if _, ok := val.(string); ok {
+			if val == "" {
+				if e.LintMode {
+					// Don't fail on missing required values when linting
+					log.Printf("[INFO] Missing required value: %s", warn)
+					return "", nil
+				}
+				return val, errors.Errorf(warn)
+			}
+		}
+		return val, nil
+	}
+
 	t.Funcs(funcMap)
 }
 

--- a/pkg/engine/engine_test.go
+++ b/pkg/engine/engine_test.go
@@ -480,6 +480,33 @@ func TestAlterFuncMap_require(t *testing.T) {
 	if gotNum := out["conan/templates/bases"]; gotNum != expectNum {
 		t.Errorf("Expected %q, got %q (%v)", expectNum, gotNum, out)
 	}
+
+	// test required without passing in needed values with lint mode on
+	// verifies lint replaces required with an empty string (should not fail)
+	lintValues := chartutil.Values{
+		"Values": chartutil.Values{
+			"who": "us",
+		},
+		"Chart": c.Metadata,
+		"Release": chartutil.Values{
+			"Name": "That 90s meme",
+		},
+	}
+	var e Engine
+	e.LintMode = true
+	out, err = e.Render(c, lintValues)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expectStr = "All your base are belong to us"
+	if gotStr := out["conan/templates/quote"]; gotStr != expectStr {
+		t.Errorf("Expected %q, got %q (%v)", expectStr, gotStr, out)
+	}
+	expectNum = "All  of them!"
+	if gotNum := out["conan/templates/bases"]; gotNum != expectNum {
+		t.Errorf("Expected %q, got %q (%v)", expectNum, gotNum, out)
+	}
 }
 
 func TestAlterFuncMap_tpl(t *testing.T) {

--- a/pkg/engine/funcs.go
+++ b/pkg/engine/funcs.go
@@ -24,7 +24,6 @@ import (
 
 	"github.com/BurntSushi/toml"
 	"github.com/Masterminds/sprig"
-	"github.com/pkg/errors"
 	yaml "gopkg.in/yaml.v2"
 )
 
@@ -54,13 +53,13 @@ func funcMap() template.FuncMap {
 		"fromYaml": fromYAML,
 		"toJson":   toJSON,
 		"fromJson": fromJSON,
-		"required": required,
 
 		// This is a placeholder for the "include" function, which is
 		// late-bound to a template. By declaring it here, we preserve the
 		// integrity of the linter.
-		"include": func(string, interface{}) string { return "not implemented" },
-		"tpl":     func(string, interface{}) interface{} { return "not implemented" },
+		"include":  func(string, interface{}) string { return "not implemented" },
+		"tpl":      func(string, interface{}) interface{} { return "not implemented" },
+		"required": func(string, interface{}) (interface{}, error) { return "not implemented", nil },
 	}
 
 	for k, v := range extra {
@@ -68,17 +67,6 @@ func funcMap() template.FuncMap {
 	}
 
 	return f
-}
-
-func required(warn string, val interface{}) (interface{}, error) {
-	if val == nil {
-		return val, errors.Errorf(warn)
-	} else if _, ok := val.(string); ok {
-		if val == "" {
-			return val, errors.Errorf(warn)
-		}
-	}
-	return val, nil
 }
 
 // toYAML takes an interface, marshals it to yaml, and returns a string. It will

--- a/pkg/engine/funcs_test.go
+++ b/pkg/engine/funcs_test.go
@@ -30,10 +30,6 @@ func TestFuncs(t *testing.T) {
 		tpl, expect string
 		vars        interface{}
 	}{{
-		tpl:    `All {{ required "A valid 'bases' is required" .bases }} of them!`,
-		expect: `All 2 of them!`,
-		vars:   map[string]interface{}{"bases": 2},
-	}, {
 		tpl:    `{{ toYaml . }}`,
 		expect: `foo: bar`,
 		vars:   map[string]interface{}{"foo": "bar"},

--- a/pkg/lint/rules/template.go
+++ b/pkg/lint/rules/template.go
@@ -68,6 +68,7 @@ func Templates(linter *support.Linter, values map[string]interface{}, namespace 
 	}
 	var e engine.Engine
 	e.Strict = strict
+	e.LintMode = true
 	renderedContentMap, err := e.Render(chart, valuesToRender)
 
 	renderOk := linter.RunLinterRule(support.ErrorSev, path, err)

